### ISSUE TITLE
Video autoplay having issue with preroll ads 

### ISSUE THREAD

**Issue by omeryounus** [googleads/videojs-ima#443]
I have implemented videojs Autoplay which is working fine, however when the ads are running and if click on ads it will open the browser tab with ads, but start the video behind the ads container video.

I do need the click function to play the videos for inline videos, for this reason I need to have the click once function for the player..

Here is the code for the inline videos to play for Mobile as well....

     	this.one("click", function() {
			    if (navigator.userAgent.match(/iPhone/i) ||
					  navigator.userAgent.match(/iPad/i) ||
					  navigator.userAgent.match(/Android/i)){
				
					console.log('mob startEvent');
			    	this.ima.initializeAdDisplayContainer();
			    	console.log('startEvent initializeAdDisplayContainer');
			    	this.ima.requestAds();
			    	console.log('mob startEvent requestAds');
				
			    } 
			    	this.play();       // this is the problem, when ads are showing this method gets called on click
			});
			

Really appreciate if someone can help me on this issue....

**Comment by shawnbuso** [comment]
Hey there,

Have you taken a look at our [autoplay sample](https://github.com/googleads/videojs-ima/tree/master/examples/autoplay)? I'm a bit confused by your use case - if you have autoplay set up, why do you need a click handler to start ads and content? Aren't they playing automatically?

Thanks,
Shawn

**Comment by omeryounus** [comment]
Hey Shawn,
Thanks for your reply....

Basically, I need the click listener for inline article videos ( which are not autoplay and click by user to play) and also on mobile browser to play the video.

The only problem that I'm facing is with autoplay video, when autoplay videos start it request for ima ads and if click on this ad container it send the click event back to videoplayer.one("click") method which play the video in background. How can we avoid this situation?

Below is the link for you to understand the situation.. as you will see video autoplay and load the ads if click on ad it will start to play it. However for quick fix I'm checking if not video playing then play but it's not always work.

https://goo.gl/j152h8

Please advise if there is any better way to deal with this issue...

**Comment by shawnbuso** [comment]
I'm not seeing ads on that page, I just go straight to the content video, so I can't repro :/

Are you adding that click handler to play the content, or is the player doing that on its own? If you're adding it can you just remove it when you autoplay? If it's the player doing it, I'd recommend taking a look at our autoplay sample (linked above) to see how you can better handle autoplay.

Would it be possible for you to send me a sample page to test with that has unobfuscated code so I can see your implementation?

**Comment by omeryounus** [comment]
Thanks for your reply....

Let me clear some confusion here: 

I have two scenarios for video to play, one is autpplay and other one is requiring the user to play the video (which is working fine on desktop/mobile etc without any issue).  The only issue is with autoplay videos, I just pass autoplay attribute in video player tag. In JavaScript initialization, I used the method below:

I know for autoplay vidoes this method not required, but since I am using single JS file so I have to use it.

     this.one(startEvent, function() {    // ideally if the videoplayer has autoplay attribute, this event should not be fire... 
      this.ima.initializeAdDisplayContainer();
      this.ima.requestAds();
      this.play();          // another solution I tried is to check if player is play/playing using the playerListenser but that didn't work.  Because sometime user before these event gets fired
}

Also, I checked your autoplay link which you are pointing too.. however, I'm having the same issue since I am not able to play video on user click. 

Basically, what I need is to handle the autoplay and click to play video using the same JavaScript initialization. So the only thing, that I need is to detect in JS 
  (player.one('click', function(){ 
    // check here if player is autoplaying, this check should be fastest or quickest way possible.
      if (!true)
        player.play();        
      }
)
Regarding the ads, I am able to see the ads, maybe some ad blocker or geolocation issue on your side. please refresh the page maybe it will appear.

**Comment by shawnbuso** [comment]
OK that makes sense - in that case I think you should be able to just listen for the 'play' event from the player to detect that it has started. Something like this:

```
player.one('play', function() {
  this.playing = true;
});

this.one(startEvent, function() {
  if (!this.playing) {
    this.ima.initializeAdDisplayContainer();
    this.ima.requestAds();
    this.play(); 
  }
});
```

Does that work?

**Comment by omeryounus** [comment]
Yes I tried this and it is currently running, but for some reason it didn't work always.

**Comment by shawnbuso** [comment]
Hmm that should be working - can you add logging to see if the play event is fired before the ad starts?

**Comment by omeryounus** [comment]
yes that play event is fired before the ads event, In play method  I am calling other method which I believe taking sometime and not setting the playing (var) property to true.. Do you think it is advisable to use another play event: 

player.one('play', function() {
  this.playing = true;
});
      current code which we use:
		var playing = false;
                this.on('play',function() { $('.video-span-caption').hide(); playing= true; 
               if(!varMap['isPlayed'+idPostFix]){ varMap['isPlayed'+idPostFix]=true; ga_track("video", 'plays', 
                videoLable); }  
               });

**Comment by shawnbuso** [comment]
I'm not sure if the jQuery hide is async or not - you could try making `playing=true` the first line of your event handler to see if that helps.

**Comment by shawnbuso** [comment]
Closing since we haven't heard from you in a bit - if you're still seeing the issue feel free to reach out and we'll be happy to look into it with you.
